### PR TITLE
Fixes #2317 predict api not working with asymmetric models

### LINKED ISSUE #2317: [BUG] `_predict` API not working with asymmetric embedding models
**What is the bug?**
Predict API not working with asymmetric models.

**How can one reproduce the bug?**
1. create a model group
2. register an asymmetric model (cf. https://github.com/opensearch-project/ml-commons/pull/2123#issue-2138948081)
3. run inference with it, using the correct parameter
4. an error states that the parameter was not passed.

**What is the expected behavior?**
The prediction should work.

**What is your host/environment?**
 - OS: MacOS
 - Version Sonoma
 - Plugins

**Do you have any screenshots?**
-
**Do you have any additional context?**
I have a fix already.

### PR DESCRIPTION
**Pull request by br3no**
### Description
When using the `_predict` API for text embeddings, the instantiation of the `MLInput` object happens via the MLCommonsClassLoader here: https://github.com/opensearch-project/ml-commons/blob/79b9dfd7eeef051fbf9d0d35242052433b83b297/common/src/main/java/org/opensearch/ml/common/input/MLInput.java#L214

The reflection logic ends up calling the constructor of `TextDocsMLInput`, which does not parse any parameters that might be contained in the `XContentParser`, though. This causes the error reported in #2317. This PR adds a case to the parsing logic to make sure the parameters are parsed. It replicates the behavior of the `MLInput` `parse` method: https://github.com/opensearch-project/ml-commons/blob/79b9dfd7eeef051fbf9d0d35242052433b83b297/common/src/main/java/org/opensearch/ml/common/input/MLInput.java#L239
 
### Issues Resolved
#2317
 
### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dhrubo-os** [comment]
> It would be nice to add an integration test for this, but I couldn't find one for the text embeddings I could easily extend.

May be we can get some idea from [here](https://github.com/opensearch-project/ml-commons/blob/fea98e050c3aca62aa64c91db80176b7d6fcdf78/plugin/src/test/java/org/opensearch/ml/rest/RestMLCustomModelActionIT.java#L35)?

File: common/src/main/java/org/opensearch/ml/common/input/nlp/TextDocsMLInput.java [hunk 1]
```diff
@@ -93,6 +96,10 @@ public TextDocsMLInput(XContentParser parser, FunctionName functionName) throws
             parser.nextToken();
 
             switch (fieldName) {
+                case ML_PARAMETERS_FIELD:
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/input/nlp/TextDocsMLInput.java:99]: Can we add corresponding unit test?

Comment (br3no) [reply common/src/main/java/org/opensearch/ml/common/input/nlp/TextDocsMLInput.java:99]: Done.

**Comment by br3no** [comment]
> > It would be nice to add an integration test for this, but I couldn't find one for the text embeddings I could easily extend.
> 
> May be we can get some idea from [here](https://github.com/opensearch-project/ml-commons/blob/fea98e050c3aca62aa64c91db80176b7d6fcdf78/plugin/src/test/java/org/opensearch/ml/rest/RestMLCustomModelActionIT.java#L35)?

This seems like a good start, but the test is annotated with `@Ignore`. I'll see if this works nonetheless.

**Comment by br3no** [comment]
The IT seems to block somewhere. I see the following comment in the [commit](https://github.com/opensearch-project/ml-commons/commit/ce6e6c74a6e0dc687aeeabe646482616effd8534) where the `@Ignore` was added:
> * Ignore problematic IT to unblock build pipeline

**Comment by br3no** [comment]
I'd like to keep this PR small and not design a new IT for embedding models, as I don't have the time now to do it.

**Comment by ylwu-amzn** [comment]
Can you share some sample requests/response to verify the fix can work?

**Comment by br3no** [comment]
> Can you share some sample requests/response to verify the fix can work?

@ylwu-amzn Sure! 

Assuming you have a cluster running that allows uploading models and can run ml-commons:
(using httpie in the examples below)

1. Create a model group:
```bash
http POST localhost:9200/_plugins/_ml/model_groups/_register <<< '                                                                                                      
{                
  "name": "local_model_group",   
  "description": "A model group for local models"
}'
```

2. Register a model:
```bash
http POST localhost:9200/_plugins/_ml/models/_register <<< '
{
    "name": "e5-small-onnx-test",
    "version": "1.0.0",
    "description": "test model",
    "model_format": "ONNX",
    "model_group_id": "GF7U0Y4B9r_1IumGvZGI",
    "model_content_hash_value": "f0ee1fce47e73bc8ae1343b74856ae63b06f2b661ebd09d9e588fff5429764ed",
    "model_config": {
        "model_type": "bert",
        "embedding_dimension": 384,
        "framework_type": "sentence_transformers",
        "query_prefix" : "query: ",
        "passage_prefix" : "passage: ",
        "all_config" : "{  \"_name_or_path\": \"intfloat/multilingual-e5-small\",  \"architectures\": [    \"BertModel\"  ],  \"attention_probs_dropout_prob\": 0.1,  \"classifier_dropout\": null,  \"hidden_act\": \"gelu\",  \"hidden_dropout_prob\": 0.1,  \"hidden_size\": 384,  \"initializer_range\": 0.02,  \"intermediate_size\": 1536,  \"layer_norm_eps\": 1e-12,  \"max_position_embeddings\": 512,  \"model_type\": \"bert\",  \"num_attention_heads\": 12,  \"num_hidden_layers\": 12,  \"pad_token_id\": 0,  \"position_embedding_type\": \"absolute\",  \"tokenizer_class\": \"XLMRobertaTokenizer\",  \"transformers_version\": \"4.30.2\",  \"type_vocab_size\": 2,  \"use_cache\": true,  \"vocab_size\": 250037}"
    },
    "url": "http://localhost:8000/intfloat-multilingual-e5-small-onnx.zip"
}
```
Replace the model group id and other parameters according to the model you want to upload. You can use any model for that, no need to use one from the e5 family. **The important thing is to make sure you have at least one of `passage_prefix` and `query_prefix`**.

3. Deploy the model
4. Call the inference endpoint with the asymmetric model parameters:
```bash
http POST localhost:9200/_plugins/_ml/_predict/text_embedding/HF7V0Y4B9r_1IumGpZGP <<< '
{
  "parameters" : {
    "content_type" : "passage"
  },
  "text_docs":[ "today is sunny"], "target_response": ["sentence_embedding"]
}'
```
5. Make sure you have embeddings as a result. You can then repeat the inference call changing the parameter to `query` and verify that the embedding values change.

**Comment by dhrubo-os** [comment]
> > Can you share some sample requests/response to verify the fix can work?
> 
> @ylwu-amzn Sure!
> 
> Assuming you have a cluster running that allows uploading models and can run ml-commons: (using httpie in the examples below)
> 
>     1. Create a model group:
> 
> 
> ```shell
> http POST localhost:9200/_plugins/_ml/model_groups/_register <<< '                                                                                                      
> {                
>   "name": "local_model_group",   
>   "description": "A model group for local models"
> }'
> ```
> 
>     2. Register a model:
> 
> 
> ```shell
> http POST localhost:9200/_plugins/_ml/models/_register <<< '
> {
>     "name": "e5-small-onnx-test",
>     "version": "1.0.0",
>     "description": "test model",
>     "model_format": "ONNX",
>     "model_group_id": "GF7U0Y4B9r_1IumGvZGI",
>     "model_content_hash_value": "f0ee1fce47e73bc8ae1343b74856ae63b06f2b661ebd09d9e588fff5429764ed",
>     "model_config": {
>         "model_type": "bert",
>         "embedding_dimension": 384,
>         "framework_type": "sentence_transformers",
>         "query_prefix" : "query: ",
>         "passage_prefix" : "passage: ",
>         "all_config" : "{  \"_name_or_path\": \"intfloat/multilingual-e5-small\",  \"architectures\": [    \"BertModel\"  ],  \"attention_probs_dropout_prob\": 0.1,  \"classifier_dropout\": null,  \"hidden_act\": \"gelu\",  \"hidden_dropout_prob\": 0.1,  \"hidden_size\": 384,  \"initializer_range\": 0.02,  \"intermediate_size\": 1536,  \"layer_norm_eps\": 1e-12,  \"max_position_embeddings\": 512,  \"model_type\": \"bert\",  \"num_attention_heads\": 12,  \"num_hidden_layers\": 12,  \"pad_token_id\": 0,  \"position_embedding_type\": \"absolute\",  \"tokenizer_class\": \"XLMRobertaTokenizer\",  \"transformers_version\": \"4.30.2\",  \"type_vocab_size\": 2,  \"use_cache\": true,  \"vocab_size\": 250037}"
>     },
>     "url": "http://localhost:8000/intfloat-multilingual-e5-small-onnx.zip"
> }
> ```
> 
> Replace the model group id and other parameters according to the model you want to upload. You can use any model for that, no need to use one from the e5 family. **The important thing is to make sure you have at least one of `passage_prefix` and `query_prefix`**.
> 
>     3. Deploy the model
> 
>     4. Call the inference endpoint with the asymmetric model parameters:
> 
> 
> ```shell
> http POST localhost:9200/_plugins/_ml/_predict/text_embedding/HF7V0Y4B9r_1IumGpZGP <<< '
> {
>   "parameters" : {
>     "content_type" : "passage"
>   },
>   "text_docs":[ "today is sunny"], "target_response": ["sentence_embedding"]
> }'
> ```
> 
>     5. Make sure you have embeddings as a result. You can then repeat the inference call changing the parameter to `query` and verify that the embedding values change.

@br3no Did you test the fix in multi-node cluster? If not could you please do that?

**Comment by br3no** [comment]
> @br3no Did you test the fix in multi-node cluster? If not could you please do that?

@dhrubo-os this was a very good hint. 

Testing in the multi-node cluster setup I ran into some very unexpected class not found errors. It turns out all `MLAlgoParameter`s must live in the "org.opensearch.ml.common.input.parameter" package. Cf. https://github.com/opensearch-project/ml-commons/blob/3cc2fcaa90f37e1eec6f5f7ba2599bdc88f1b637/common/src/main/java/org/opensearch/ml/common/MLCommonsClassLoader.java#L87

This is very obscure and I can't say why the errors only appeared in a multi-node setting. But now it works.

Let me know if there is anything else open still.

**Review by dhrubo-os — APPROVED** [review]

**Comment by dhrubo-os** [comment]
> > @br3no Did you test the fix in multi-node cluster? If not could you please do that?
> 
> @dhrubo-os this was a very good hint.
> 
> Testing in the multi-node cluster setup I ran into some very unexpected class not found errors. It turns out all `MLAlgoParameter`s must live in the "org.opensearch.ml.common.input.parameter" package. Cf.
> 
> https://github.com/opensearch-project/ml-commons/blob/3cc2fcaa90f37e1eec6f5f7ba2599bdc88f1b637/common/src/main/java/org/opensearch/ml/common/MLCommonsClassLoader.java#L87
> 
> This is very obscure and I can't say why the errors only appeared in a multi-node setting. But now it works.
> 
> Let me know if there is anything else open still.

Thanks for looking into this. Approved.

**Comment by br3no** [comment]
@ylwu-amzn mind having a look? Only one review required to merge this...

@dhrubo-os, or maybe there is someone else available?

**Review by ylwu-amzn — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**